salt/modules/opkg.py

```diff
@@ -0,0 +1,975 @@
+    versions.
+    '''
+    ret = {}
+    valid_states = ('hold', 'noprune', 'user', 'ok', 'installed', 'unpacked')
+    if state not in valid_states:
+        raise SaltInvocationError(
+            'Invalid state: {0}'.format(state)
+        )
+    oldstate = _get_state(pkg)
+    cmd = ['opkg', 'flag']
+    cmd.append(state)
+    cmd.append(pkg)
+    _out = __salt__['cmd.run'](cmd)
+
+    # Missing return value check due to opkg issue 160
+    ret[pkg] = {'old': oldstate,
+                'new': state}
+    return ret
+
+
+def list_pkgs(versions_as_list=False, **kwargs):
+    '''
+    List the packages currently installed in a dict::
+
+        {'<package_name>': '<version>'}
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' pkg.list_pkgs
+        salt '*' pkg.list_pkgs versions_as_list=True
+    '''
+    versions_as_list = salt.utils.is_true(versions_as_list)
+    # not yet implemented or not applicable
+    if any([salt.utils.is_true(kwargs.get(x))
+            for x in ('removed', 'purge_desired')]):
+        return {}
+
+    if 'pkg.list_pkgs' in __context__:
+        if versions_as_list:
+            return __context__['pkg.list_pkgs']
+        else:
+            ret = copy.deepcopy(__context__['pkg.list_pkgs'])
+            __salt__['pkg_resource.stringify'](ret)
+            return ret
+
+    cmd = ['opkg', 'list-installed']
+    ret = {}
+    out = __salt__['cmd.run'](cmd, output_loglevel='trace')
+    for line in out.splitlines():
+        pkg_name, pkg_version = line.split(' - ')
+        __salt__['pkg_resource.add_pkg'](ret, pkg_name, pkg_version)
+
+    __salt__['pkg_resource.sort_pkglist'](ret)
+    __context__['pkg.list_pkgs'] = copy.deepcopy(ret)
+    if not versions_as_list:
+        __salt__['pkg_resource.stringify'](ret)
+    return ret
+
+
+def list_upgrades(refresh=True):
+    '''
+    List all available package upgrades.
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' pkg.list_upgrades
+    '''
+    ret = {}
+    if salt.utils.is_true(refresh):
+        refresh_db()
+
+    cmd = ['opkg', 'list-upgradable']
+    call = __salt__['cmd.run_all'](cmd, output_loglevel='trace')
+
+    if call['retcode'] != 0:
+        comment = ''
+        if 'stderr' in call:
+            comment += call['stderr']
+        if 'stdout' in call:
+            comment += call['stdout']
+        raise CommandExecutionError(
+                '{0}'.format(comment)
+        )
+    else:
+        out = call['stdout']
+
+    for line in out.splitlines():
+        name, _oldversion, newversion = line.split(' - ')
+        ret[name] = newversion
+
+    return ret
+
+
+def upgrade_available(name):
+    '''
+    Check whether or not an upgrade is available for a given package
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' pkg.upgrade_available <package name>
+    '''
+    return latest_version(name) != ''
+
+
+def version_cmp(pkg1, pkg2):
```

Comment:
Rather than a `try..except Exception` could you rewrite this function:
```python
import salt.utils.decorators.which as _which
...
@which('opkg-compare-versions')
def version_cmp(pkg1, pkg2):
```